crates/biome_js_type_info/src/flattening/intersections.rs

```diff
@@ -0,0 +1,319 @@
+use biome_rowan::Text;
+
+use crate::{
+    Class, Function, Interface, Intersection, Literal, Namespace, Object, ResolvedTypeData,
+    ReturnType, TypeData, TypeMember, TypeResolver,
+};
+
+pub(super) fn flattened_intersection(
+    intersection: &Intersection,
+    resolver: &mut dyn TypeResolver,
+) -> TypeData {
+    match intersection.types() {
+        [] => TypeData::NeverKeyword,
+        [ty] => resolver
+            .resolve_and_get(ty)
+            .map_or_else(TypeData::unknown, ResolvedTypeData::to_data),
+        types => {
+            let merged_ty = resolver
+                .resolve_and_get(&types[0])
+                .map_or_else(TypeData::unknown, ResolvedTypeData::to_data);
+            let mut merged_ty = MergedType::from_type(merged_ty, resolver);
+            let mut primitive = merged_ty.as_primitive().cloned();
+            for other in &types[1..] {
+                if !merged_ty.is_mergeable() {
+                    break;
+                }
+
+                let other_ty = resolver
+                    .resolve_and_get(other)
+                    .map_or_else(TypeData::unknown, ResolvedTypeData::to_data);
+                let other_ty = MergedType::from_type(other_ty, resolver);
+                if let Some(other_primitve) = other_ty.as_primitive() {
+                    if primitive.is_some() {
+                        return TypeData::NeverKeyword;
+                    }
+
+                    primitive = Some(other_primitve.clone());
+                } else {
+                    merged_ty = merged_ty.intersection_with(other_ty, resolver);
+                }
+            }
+
+            if let Some(primitive) = primitive {
+                primitive
+            } else {
+                merged_ty.into_type()
+            }
+        }
+    }
+}
+
+enum MergedType {
+    Any,
+    Class(Vec<TypeMember>),
+    Function(Box<Function>),
+    Interface(Vec<TypeMember>),
+    Namespace(Vec<TypeMember>),
+    Never,
+    Object(Vec<TypeMember>),
+    Primitive(TypeData),
+    Unknown,
+}
+
+enum MergedTypeKind {
+    Any,
+    Class,
+    Function,
+    Interface,
+    Namespace,
+    Never,
+    Object,
+    Primitive,
+    Unknown,
+}
+
+impl MergedTypeKind {
+    fn intersection_with(self, other: Self) -> Self {
```

Comment:
Can we document the business logic of this function somewhere? Either doc comments, or a readme